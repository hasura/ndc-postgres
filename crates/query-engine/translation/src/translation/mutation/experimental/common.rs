//! Some common helper functions.

use std::collections::BTreeSet;

/// Create a description string for keys. For example:
/// > "'TrackId' key", or
/// > "'TrackId', 'PlaylistId' and 'OtherId' keys".
pub fn description_keys(keys: &BTreeSet<String>) -> String {
    let mut string = String::new();

    // Case we have only one key.
    #[allow(clippy::comparison_chain)]
    if keys.len() == 1 {
        for key in keys {
            string.push_str(format!("'{key}'").as_str());
        }
        string.push_str(" key");
    // Otherwise (we don't expect to be called without any keys).
    } else {
        for (index, key) in keys.iter().enumerate() {
            string.push_str(format!("'{key}'").as_str());
            // Push a comma if we haven't reached the key before last.
            if index + 2 < keys.len() {
                string.push_str(", ");
            }
            // Push an "and" if we reached the key before last.
            else if index + 2 == keys.len() {
                string.push_str(" and ");
            }
        }
        string.push_str(" keys");
    }

    string
}
