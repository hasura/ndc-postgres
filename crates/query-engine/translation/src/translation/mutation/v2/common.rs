//! Some common helper functions.

use crate::translation::error::Warning;
use ndc_models as models;
use nonempty::NonEmpty;
use query_engine_metadata::metadata;
use std::collections::BTreeSet;

/// Create a description string for keys. For example:
/// > "'TrackId' key", or
/// > "'TrackId', 'PlaylistId' and 'OtherId' keys".
pub fn description_keys(keys: &BTreeSet<&models::FieldName>) -> String {
    let mut string = String::new();

    // Case we have only one key.
    #[allow(clippy::comparison_chain)]
    if keys.len() == 1 {
        for key in keys {
            string.push_str(format!("'{key}'").as_str());
        }
        string.push_str(" key");
    // Otherwise (we don't expect to be called without any keys).
    } else {
        for (index, key) in keys.iter().enumerate() {
            string.push_str(format!("'{key}'").as_str());
            // Push a comma if we haven't reached the key before last.
            if index + 2 < keys.len() {
                string.push_str(", ");
            }
            // Push an "and" if we reached the key before last.
            else if index + 2 == keys.len() {
                string.push_str(" and ");
            }
        }
        string.push_str(" keys");
    }

    string
}

/// Fetch the key columns of a constraint and calculate the constraint name.
pub fn get_unique_constraint_name_and_key_columns(
    mutation_type: &str,
    collection_name: &models::CollectionName,
    db_constraint_name: &str,
    table_info: &metadata::database::TableInfo,
    keys: &metadata::UniquenessConstraint,
) -> Option<(String, NonEmpty<metadata::database::ColumnInfo>)> {
    let mut key_columns: Vec<metadata::database::ColumnInfo> = vec![];
    let mut constraint_name = String::new();

    for (index, (_, field_key)) in keys.0.iter().enumerate() {
        // We don't expect this to happen because the metadata generated should be consistent,
        // but if it does, we skip generating these procedure rather than not start at all.
        let key_column = {
            if let Some(key_column) = table_info.columns.get(field_key) {
                key_column
            } else {
                let warning = Warning::GeneratingMutationSkippedBecauseColumnNotFoundInCollection {
                    mutation_type: mutation_type.to_string(),
                    collection: collection_name.clone(),
                    column: field_key.clone(),
                    db_constraint_name: db_constraint_name.to_string(),
                };
                tracing::warn!(
                    info = ?warning,
                    warning = format!("{warning}"),
                );
                None?
            }
        };

        key_columns.push(key_column.clone());

        constraint_name.push_str(field_key.as_str());
        if index + 1 < keys.0.len() {
            constraint_name.push_str("_and_");
        }
    }
    // There should be at least one key column.
    let key_columns = if let Some(key_columns) = NonEmpty::from_vec(key_columns) {
        key_columns
    } else {
        let warning = Warning::GeneratingMutationSkippedBecauseNoColumnsInConstraint {
            mutation_type: mutation_type.to_string(),
            collection: collection_name.clone(),
            db_constraint_name: db_constraint_name.to_string(),
        };
        tracing::warn!(
            info = ?warning,
            warning = format!("{warning}"),
        );
        None?
    };

    Some((constraint_name, key_columns))
}

/// The name and description of a check input argument.
#[derive(Debug, Clone)]
pub struct CheckArgument {
    pub argument_name: models::ArgumentName,
    pub description: String,
}

// Default check argument/constraint
pub fn default_constraint() -> serde_json::Value {
    serde_json::json!({"type": "and", "expressions": []})
}

// the old default was to prefix generated mutations with `v2_` or `v1_`
// but now we are able to override this
pub fn get_version_prefix(mutations_prefix: Option<&String>) -> String {
    match mutations_prefix {
        None => format!("{}_", super::VERSION),
        Some(str) => match str.as_str() {
            "" => String::new(),
            _ => format!("{str}_"),
        },
    }
}

#[test]
fn test_version_prefix() {
    assert_eq!(get_version_prefix(None), "v2_".to_string());
    assert_eq!(
        get_version_prefix(Some("horse".into()).as_ref()),
        "horse_".to_string()
    );
    assert_eq!(
        get_version_prefix(Some(String::new()).as_ref()),
        String::new()
    );
}
