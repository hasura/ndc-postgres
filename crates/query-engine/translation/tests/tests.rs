mod common;

#[test]
fn select_where_album_id_equals_self_nested_object_relationship() {
    let result =
        common::test_translation("select_where_album_id_equals_self_nested_object_relationship")
            .unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_converts_select_with_limit() {
    let result = common::test_translation("select_with_limit").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_select_where_string() {
    let result = common::test_translation("select_where_string").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_select_where_not_null() {
    let result = common::test_translation("select_where_not_null").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_select_where_unrelated_exists() {
    let result = common::test_translation("select_where_unrelated_exists").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_select_where_related_exists() {
    let result = common::test_translation("select_where_related_exists").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_aggregate_count_albums() {
    let result = common::test_translation("aggregate_count_albums").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_aggregate_distinct_albums() {
    let result = common::test_translation("aggregate_distinct_albums").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_aggregate_function_albums() {
    let result = common::test_translation("aggregate_function_albums").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_simple_array_relationship() {
    let result = common::test_translation("simple_array_relationship").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn it_simple_object_relationship() {
    let result = common::test_translation("simple_object_relationship").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn nested_array_relationships() {
    let result = common::test_translation("nested_array_relationships").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn nested_aggregates() {
    let result = common::test_translation("nested_aggregates").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn dup_array_relationship() {
    let result = common::test_translation("dup_array_relationship").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn nested_recursive_relationship() {
    let result = common::test_translation("nested_recursive_relationship").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn very_nested_recursive_relationship() {
    let result = common::test_translation("very_nested_recursive_relationship").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn sorting_by_relationship_column() {
    let result = common::test_translation("sorting_by_relationship_column").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn sorting_by_nested_relationship_column() {
    let result = common::test_translation("sorting_by_nested_relationship_column").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn sorting_by_recursive_relationship_column() {
    let result = common::test_translation("sorting_by_recursive_relationship_column").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn sorting_by_relationship_count() {
    let result = common::test_translation("sorting_by_relationship_count").unwrap();
    insta::assert_snapshot!(result);
}

#[test]
fn sorting_by_nested_relationship_count() {
    let result = common::test_translation("sorting_by_nested_relationship_count").unwrap();
    insta::assert_snapshot!(result);
}

mod native_queries {
    use crate::common;

    #[test]
    fn select_artist() {
        let result = common::test_translation("native_queries/select_artist").unwrap();
        insta::assert_snapshot!(result);
    }

    #[test]
    fn select_artist_by_id() {
        let result = common::test_translation("native_queries/select_artist_by_id").unwrap();
        insta::assert_snapshot!(result);
    }

    #[test]
    fn select_artist_by_name() {
        let result = common::test_translation("native_queries/select_artist_by_name").unwrap();
        insta::assert_snapshot!(result);
    }

    #[test]
    fn select_artist_with_album_by_title() {
        let result =
            common::test_translation("native_queries/select_artist_with_album_by_title").unwrap();
        insta::assert_snapshot!(result);
    }

    #[test]
    fn select_artist_with_album_by_title_relationship_arguments() {
        let result = common::test_translation(
            "native_queries/select_artist_with_album_by_title_relationship_arguments",
        )
        .unwrap();
        insta::assert_snapshot!(result);
    }
}

mod types {
    use crate::common;

    #[test]
    fn select_types_on_comparisons() {
        let result = common::test_translation("values/select_types_on_comparisons").unwrap();
        insta::assert_snapshot!(result);
    }

    #[test]
    fn select_types_on_native_queries() {
        let result = common::test_translation("values/select_types_on_native_queries").unwrap();
        insta::assert_snapshot!(result);
    }
}
