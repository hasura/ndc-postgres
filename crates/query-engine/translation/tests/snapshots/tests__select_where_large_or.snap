---
source: crates/query-engine/translation/tests/tests.rs
assertion_line: 339
expression: result
snapshot_kind: text
---
SELECT
  coalesce(json_agg(row_to_json("%2_universe")), '[]') AS "universe"
FROM
  (
    SELECT
      *
    FROM
      (
        SELECT
          coalesce(json_agg(row_to_json("%3_rows")), '[]') AS "rows"
        FROM
          (
            SELECT
              "%1_Album"."AlbumId" AS "AlbumId"
            FROM
              (
                SELECT
                  "%0_Album".*
                FROM
                  "public"."Album" AS "%0_Album"
                WHERE
                  (
                    "%0_Album"."Title" IN (
                      cast($1 as "pg_catalog"."varchar"),
                      cast($2 as "pg_catalog"."varchar"),
                      cast($3 as "pg_catalog"."varchar"),
                      cast($4 as "pg_catalog"."varchar"),
                      cast($5 as "pg_catalog"."varchar"),
                      cast($6 as "pg_catalog"."varchar"),
                      cast($7 as "pg_catalog"."varchar"),
                      cast($8 as "pg_catalog"."varchar"),
                      cast($9 as "pg_catalog"."varchar"),
                      cast($10 as "pg_catalog"."varchar")
                    )
                  )
              ) AS "%1_Album"
          ) AS "%3_rows"
      ) AS "%3_rows"
  ) AS "%2_universe";

{
    1: String(
        "Album1",
    ),
    2: String(
        "Album2",
    ),
    3: String(
        "Album3",
    ),
    4: String(
        "Album4",
    ),
    5: String(
        "Album5",
    ),
    6: String(
        "Album6",
    ),
    7: String(
        "Album7",
    ),
    8: String(
        "Album8",
    ),
    9: String(
        "Album9",
    ),
    10: String(
        "Album10",
    ),
}
